scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/language/json/JsonTemplateParser.scala

```diff
@@ -0,0 +1,181 @@
+        .map(_.mkString)
+    case expression: LiteralExpression => validNel(expression.getValue)
+    case expression: standard.SpelExpression =>
+      spelExpressionToDefaultJsonValue(expression.getExpressionString, Typed[SpelExpressionRepr])
+    case _ => invalidNel(ExpressionCompilationError("Unknown expression type"))
+  }
+
+  private def spelExpressionToDefaultJsonValue(ctx: ValidationContext)(
+      original: String,
+      expectedType: typing.TypingResult
+  ): ValidatedNel[ExpressionParseError, String] = spelParser
+    .parse(original, ctx, expectedType)
+    .map(_.typingInfo.typingResult)
+    .map(typingResultToDefaultJsonValue)
+
+  private def typingResultToDefaultJsonValue(typingResult: TypingResult): String = typingResult match {
+    case TypedClass(k, _) =>
+      k match {
+        case className if TypeValueDeterminer.isIntegerNumber(className)       => "0"
+        case className if TypeValueDeterminer.isFloatingPointNumber(className) => "0.5"
+        case className if TypeValueDeterminer.isBoolean(className)             => "true"
+        case className if TypeValueDeterminer.isString(className)              => "unquoted string"
+        // For now, complex types are treated as String
+        case _ => "unquoted string"
+      }
+    // For now, complex types are treated as String
+    case _ => "unquoted string"
+  }
+
+  private def extractJsonStringFromCompiledExpression(
+      compiledExpression: CompiledExpression
+  ): ValidatedNel[ExpressionParseError, String] =
+    compiledExpression match {
+      case expression: SpelExpression =>
+        joinLiteralsAndReplaceExpressionWithDefaults(
+          spelExpressionToDefaultJsonValue,
+          expression.parsedSpringExpression,
+        )
+      case _ => invalidNel(ExpressionCompilationError("Invalid compiled expression type"))
+    }
+
+  private def spelExpressionToDefaultJsonValue(
+      original: String,
+      expectedType: typing.TypingResult
+  ): ValidatedNel[ExpressionParseError, String] = spelParser
+    .parseWithoutContextValidation(original, expectedType)
+    .map(_ => "0")
+
+}
+
+object JsonTemplateParser {
+  private val stringTypingResult = Typed.typedClass[String]
+
+  case class CompiledJsonTemplateExpression(
+      languageId: Language,
+      originalJsonString: String,
+      templateCompiledExpression: CompiledExpression,
+      expectedType: typing.TypingResult
+  ) extends CompiledExpression {
+
+    override def language: Language = languageId
+
+    override def original: String = originalJsonString
+
+    override def evaluate[T](ctx: Context, globals: Map[String, Any]): T =
+      if (expectedType == stringTypingResult) {
+        templateCompiledExpression.evaluate[T](ctx, globals)
+      } else {
+        val jsonString = templateCompiledExpression.evaluate[String](ctx, globals)
+        parser.parse(jsonString) match {
+          case Left(error)  => throw new IllegalStateException("Parsing JSON failed with error", error)
```

Comment:
`IllegalStateException` means that we didn't expect something. We encounter this situation when, e.g. someone will put something to a placeholder (for example list), which will break the json. We know that it may happen, and we accept it. I think that it will be better to throw our own exception with error as a cause or just error which is also an exception (see `ParsingFailure`)